### DISCUSSION TITLE
Get a copy of the raw input from inside model validator

### DISCUSSION THREAD

**Discussion by shadycuz** [pydantic/pydantic#7266] [General]
I have a function that does model validation:

`@model_validator(mode="after")`

At the end of the validation I would like to save a copy of the raw user input used to create the model but I can't seem to find a way. to access it. Most of the functions like model_dump contain the "computed" fields that could have been manipulated by field validators. 

For example we have several fields like `app_version` and `app_build` which if set to `latest` a field validator will replace `latest` with the correct version string.

**Comment by shadycuz** [comment]
I tried using a private attribute like:

```
_raw_input: Optional[Dict[str, Any]] = PrivateAttr(default=None)

    def __init__(self, **data):
        """Initialize the StackConfig object.

        Args:
            **data (Any): The data to initialize the object with.
        """
        super().__init__(**data)
        self._raw_input = data
```

But to my surprise a `@model_validator(mode="after")` actually runs BEFORE the `__init__` function. So that didn't work either.

**Comment by bacala** [comment]
Hi shadycuz,

You probably should use \_\_new__() instead of \_\_init__(). It runs the very first.

**Reply by shadycuz** [reply]
@bacala using __new__ results in this Error:

> AttributeError: 'StackConfig' object has no attribute '__pydantic_private__'. Did you mean: '__pydantic_complete__'?

Apparently the "PrivateAttr" are created in __init__, so you cant set them until __init__ is called. See https://github.com/PromtEngineer/localGPT/issues/672

This seems to be working

```
@model_validator(mode="before")
@classmethod
def save_raw_input(cls, data: Any) -> Any:

    cls._raw_input = data
    return data
```

**Reply by naquiroz** [reply]
This is not thread-safe. If instantiating two instances of the model at the same time, this could lead to inconsistencies.

**Reply by shadycuz** [reply]
It also seemed to stopped working when I added `model_config = ConfigDict(validate_assignment=True)`

So I'm back to square one when it comes to saving the initial state of the model.

**Comment by danielquinn** [comment]
This seems to be doing the job for me:

```python
from typing import Any, Self
from pydantic import (
    BaseModel,
    ConfigDict,
    ModelWrapValidatorHandler,
    model_validator,
)


class Alpha(BaseModel):
    x: int = 4
    _raw: Any | None = None

    @model_validator(mode="wrap")
    @classmethod
    def build_extra(
        cls,
        data: Any,
        handler: ModelWrapValidatorHandler[Self],
    ) -> Self:
        r = handler(data)
        r._raw = data
        return r

a = Alpha(x=2, y=7)
print(a._raw)
{'x': 2, 'y': 7}
```

I've tried this with and without `model_config = ConfigDict(validate_assignment=True)` and I get the same results.
